### ISSUE TITLE
pipeline callback invoked after the pipeline has already completed

### ISSUE THREAD

**Issue by lderezinski** [joyent/node-manta#228]
Sanitized the path being removed.  note it was this kind of structure /user/stor/valid/path/here/YYYY/MM/DD/(10-20 files in here)

```
 #  mrm -r  /user/stor/valid/path/here
 mrm: pipeline callback invoked after the pipeline has already completed ({ operations: 
   [ { func: [Function: createAgent],
       funcname: 'createAgent',
       status: 'ok',
       err: undefined,
       result: undefined },
     { func: [Function: checkAgentForKey],
       funcname: 'checkAgentForKey',
       status: 'ok',
       err: undefined,
       result: undefined },
     { func: [Function: loadKey],
       funcname: 'loadKey',
       status: 'ok',
       err: undefined,
       result: undefined } ],
  successes: [ undefined, undefined, undefined ],
  ndone: 3,
  nerrors: 0 })
```

**Comment by davepacheco** [comment]
Unfortunately, this has been seen only a few times and only in one environment.  It looks like an uncommon race in the signing initialization path.  The 'uncaughtException' listener that prints the error message also prevents us from getting a core file that would tell us which callback was invoked an extra time, through what stack, and with what surrounding state.  All we can do is try to reproduce ourselves or inspect the code.  I didn't find any obvious bugs in the code, though the ssh agent code is suspicious in that it expects to get all data from the agent in one 'data' event.  If that were fragmented into multiple 'data' events for some reason, that might cause this.

Short of reproducing it without the uncaughtException handler (which would require a code change), if you can reproduce it with DEBUG=1 in the environment, that may narrow down the cause considerably.

In the meantime, I've filed #229.

**Comment by lderezinski** [comment]
here is the debug output.  In this case it is from a mjob get UUID

mjob: pipeline callback invoked after the pipeline has already completed ({ operations: 
   [ { func: [Function: createAgent],
       funcname: 'createAgent',
       status: 'ok',
       err: undefined,
       result: undefined },
     { func: [Function: checkAgentForKey],
       funcname: 'checkAgentForKey',
       status: 'ok',
       err: undefined,
       result: undefined },
     { func: [Function: loadKey],
       funcname: 'loadKey',
       status: 'ok',
       err: undefined,
       result: undefined } ],
  successes: [ undefined, undefined, undefined ],
  ndone: 3,
  nerrors: 0 })
    at next (/root/scripts/node_modules/manta/node_modules/vasync/lib/vasync.js:160:10)
    at /root/scripts/node_modules/manta/lib/auth.js:453:21
    at /root/scripts/node_modules/manta/lib/auth.js:240:9
    at identitiesAnswer (/root/scripts/node_modules/manta/node_modules/ssh-agent/lib/ssh_agent_client.js:179:12)
    at Socket.<anonymous> (/root/scripts/node_modules/manta/node_modules/ssh-agent/lib/ssh_agent_client.js:279:12)
    at Socket.EventEmitter.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:746:14)
    at Socket.EventEmitter.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:408:10)
    at emitReadable (_stream_readable.js:404:5)

**Comment by davepacheco** [comment]
That stack further indicts the ssh-agent code path.  We know from the pipeline data that checkAgentForKey() returned success.  We know from the stack that we wound up calling requestIdentities(), and it's returning with success.  We can infer that the first callback also came from requestIdentities(), so it also returned success.  This really feels like the ssh agent data was split into two 'data' chunks and the code doesn't handle that.

**Comment by lderezinski** [comment]
interesting, as I only have two keys loaded

**Comment by lderezinski** [comment]
node --abort-on-uncaught-exception $(which mjob) get UUID

Uncaught pipeline callback invoked after the pipeline has already completed ({ operations: 
   [ { func: [Function: createAgent],
       funcname: 'createAgent',
       status: 'ok',
       err: undefined,
       result: undefined },
     { func: [Function: checkAgentForKey],
       funcname: 'checkAgentForKey',
       status: 'ok',
       err: undefined,
       result: undefined },
     { func: [Function: loadKey],
       funcname: 'loadKey',
       status: 'ok',
       err: undefined,
       result: undefined } ],
  successes: [ undefined, undefined, undefined ],
  ndone: 3,
  nerrors: 0 })

FROM
next (/root/scripts/node_modules/manta/node_modules/vasync/lib/vasync.js:160:10)
/root/scripts/node_modules/manta/lib/auth.js:453:21
/root/scripts/node_modules/manta/lib/auth.js:240:9
Socket.<anonymous> (/root/scripts/node_modules/manta/node_modules/ssh-agent/lib/ssh_agent_client.js:284:7)
Socket.EventEmitter.emit (events.js:95:17)
Pipe.close (net.js:465:12)

**Comment by davepacheco** [comment]
Great!  Can I get a copy of that core file?

**Comment by davepacheco** [comment]
I've been seeing a similar error occasionally for the last month or two:

```
minfo: TypeError: Not a buffer
    at Sign.sign (crypto.js:398:27)
    at Object.signWithPrivateKey [as func] (/home/dap/node-v0.10.30-sunos-x86/lib/node_modules/manta/lib/auth.js:557:33)
    at /home/dap/node-v0.10.30-sunos-x86/lib/node_modules/manta/node_modules/vasync/lib/vasync.js:191:20
    at process._tickCallback (node.js:419:13)
```

Today, it was reproducible enough to take a few laps with --abort-on-uncaught-exception, and I managed to work out was going on.  As suspected from code inspection, requestIdentities() is invoking its callback twice: once from the "timeout" path, and once from the "connect" path.  Presumably what's going on is that we're successfully connecting to the agent, but only after the timeout has expired.  I'm going to submit a pull request for node-ssh-agent to fix this issue.

I worked this out by adding logging to each of the event handlers in the lib/auth.js function as well as for all the event handlers in ssh-agent.request.  I also printed out a stack whenever the sshAgentGetKey callback is called twice, and then aborted the program to get a core file.  Here's the smoking gun:

```
createSSHAgent callback
dap: ssh agent client: timeout
Error: sshAgentGetKey callback
    at /home/dap/node-manta/lib/auth.js:452:21
    at /home/dap/node-manta/lib/auth.js:240:9
    at Socket.<anonymous> (/home/dap/node-manta/node_modules/ssh-agent/lib/ssh_agent_client.js:  304:12)
    at Socket.g (events.js:180:16)
    at Socket.emit (events.js:92:17)
    at Socket._onTimeout (net.js:326:8)
    at Timer.unrefTimeout [as ontimeout] (timers.js:427:13)
dap: ssh agent client: connect
dap: ssh agent client: error Error: write after end
    at writeAfterEnd (_stream_writable.js:132:12)
    at Socket.Writable.write (_stream_writable.js:180:5)
    at Socket.write (net.js:615:40)
    at Socket.<anonymous> (/home/dap/node-manta/node_modules/ssh-agent/lib/ssh_agent_client.js:  292:12)
    at Socket.emit (events.js:117:20)
    at Object.afterConnect [as oncomplete] (net.js:886:10)
Error: sshAgentGetKey callback
    at /home/dap/node-manta/lib/auth.js:452:21
    at /home/dap/node-manta/lib/auth.js:240:9
    at Socket.<anonymous> (/home/dap/node-manta/node_modules/ssh-agent/lib/ssh_agent_client.js:  297:12)
    at Socket.emit (events.js:95:17)
    at writeAfterEnd (_stream_writable.js:134:10)
    at Socket.Writable.write (_stream_writable.js:180:5)
    at Socket.write (net.js:615:40)
    at Socket.<anonymous> (/home/dap/node-manta/node_modules/ssh-agent/lib/ssh_agent_client.js:  292:12)
    at Socket.emit (events.js:117:20)
    at Object.afterConnect [as oncomplete] (net.js:886:10)
Abort (core dumped)
```

**Comment by davepacheco** [comment]
Just released manta@1.5.2 with this fix.  Where my previous loop would run into this after a few minutes, with the fix, I didn't see the problem at all, after waiting quite a while.  For regression testing:

```
- illumos, v0.10.40, 32-bit: "make test": PASS
- illumos, v0.10.40, 64-bit: "make test": PASS
- illumos, v0.12.7,  32-bit: "make test": PASS
- illumos, v0.12.7,  64-bit: "make test": PASS
- OS X,    v0.12.2,  64-bit: "make test": PASS
- OS X,    v0.10.28, 64-bit: "make test": PASS
```
